lib/Conversion/TritonToStructured/TritonToStructuredPass.cpp

```diff
@@ -41,13 +57,244 @@ class TritonToStructuredPass
+      if (auto ptrType =
+              dyn_cast<triton::PointerType>(tensorType.getElementType())) {
+        SmallVector<Type> tupleTypes{tensorType};
+        // Each tensor of rank k gets k values for its offsets and k values for
+        // its strides, all of which has Index type.
+        tupleTypes.append(tensorType.getRank() * 2, IndexType::get(context));
+        types = SmallVector<Type>{TupleType::get(context, tupleTypes)};
+        return success();
+      }
+      return std::nullopt;
+    });
+
+    // Case 2: Block pointers (!tt.ptr<tensor<type>> or !tt.ptr<type>)
+    converter.addConversion([context](triton::PointerType ptrType,
+                                      SmallVectorImpl<Type> &types)
+                                -> std::optional<LogicalResult> {
+      if (auto tensorType =
+              llvm::dyn_cast<RankedTensorType>(ptrType.getPointeeType())) {
+        // Block pointers
+        SmallVector<Type> tupleTypes{ptrType};
+        // Each tensor of rank k gets k values for its offsets and k values for
+        // its strides, all of which has Index type.
+        tupleTypes.append(tensorType.getRank() * 2, IndexType::get(context));
+        types = SmallVector<Type>{TupleType::get(context, tupleTypes)};
+      } else {
+        // Scalar pointers
+        // The only relevant state that can be updated in loops for scalar
+        // pointers are offset. No need to include stride here.
+        auto tupleType = TupleType::get(
+            context, SmallVector<Type>{ptrType, IndexType::get(context)});
+        types = SmallVector<Type>{tupleType};
+      }
+      return success();
+    });
+
+    // Hooks to compute the correct materialization, "argument" and "source"
+    // materialization are used when we need to convert the tuple type back to
+    // the original triton pointer type. These are used when there are ops that
+    // still need to use the original pointer type. For instance, we convert the
+    // result of tt.addptr from tt.ptr type to a tuple, but the original ptr
+    // result is still being used by another tt.load or tt.store.
+    auto materialize = [](OpBuilder &builder, Type resultType,
+                          ValueRange inputs, Location loc) {
+      return builder.create<UnrealizedConversionCastOp>(loc, resultType, inputs)
+          .getResult(0);
+    };
+
+    converter.addArgumentMaterialization(materialize);
+    converter.addSourceMaterialization(materialize);
+
+    // Compute the target materialization, given a value with the pointer type,
+    // convert that value to a tuple type.
+    converter.addTargetMaterialization(
+        [](OpBuilder &builder, TypeRange resultTypes, Value input,
+           Location loc) -> std::optional<SmallVector<Value>> {
+          return builder
+              .create<UnrealizedConversionCastOp>(loc, resultTypes, input)
+              ->getResults();
+        });
+
+    scf::populateSCFStructuralOneToNTypeConversions(converter, patterns);
```

Comment:
This definitely allows us to simplify our code by quite a bit, but I am slightly worried about one specific scenario.

Assume there is an iter arg of a scf.for is a tensor of integer; the loop body uses this iter arg to create a non-block pointer. To support scenarios like this, we may want to visit the use of the iter arg before deciding that we'd want to convert this iter arg to a structured format too. 

I believe we cannot rely on TypeConverter to support such scenarios, since type rewrite cannot be conditional. 

I understand this is non-trivial, but would it be possible for us to keep the manual rewriting framework (or if you know better ways to support such cases)?

Reply:
>would it be possible for us to keep the manual rewriting framework
 i really hope there's a better way to handle this...

Reply:
@haishanzzzz @manbearian I believe we can still support the case that you mentioned using this new type converter approach.

The changes applied to the IR in the prepass are reversible by running canonicalization again. What this means is, we can always safely apply the same transformation to *all* tensors of integers to also include the additional offsets and strides *just in case* the tensors are used as part of a pointer arithmetic sequence.

The PtrAnalysis pass can then do some analysis to determine whether a tensor of integer needs to be converted to the structured format. If a structured format is necessary for a given tensor of integer, it means that the offsets and strides included in the loop are used. Otherwise, the offsets and strides are "dead" and can be safely removed by canonicalization after PtrAnalysis finishes. This approach is similar to cases where PtrAnalysis cannot analyze pointers and have to remap back to the "original pointer" -- in such cases, the offsets and strides inserted by the prepass are also removed.

Reply:
That makes sense. Thank you for the explanation. Would you want to do that within this PR, or should we leave it for later?

Reply:
I can do another follow up PR to implement this. Do you happen to have a triton test case handy? Otherwise it's fine too, I can try generating one. 😄